conda/core/package_cache_data.py

```diff
@@ -780,146 +784,166 @@ def execute(self):
+            ]
 
-        if not context.verbose and not context.quiet and not context.json:
-            print(
-                "\nDownloading and Extracting Packages:",
-                end="\n" if IS_INTERACTIVE else " ...working...",
-            )
-        else:
-            log.debug(
-                "prepared package cache actions:\n"
-                "  cache_actions:\n"
-                "    %s\n"
-                "  extract_actions:\n"
-                "    %s\n",
-                "\n    ".join(str(ca) for ca in self.cache_actions),
-                "\n    ".join(str(ea) for ea in self.extract_actions),
-            )
+            if self._executed:
+                return
+            if not self._prepared:
+                self.prepare()
 
-        exceptions = []
-        progress_bars = {}
-        futures: list[Future] = []
-
-        cancelled_flag = False
-
-        def cancelled():
-            """
-            Used to cancel download threads.
-            """
-            nonlocal cancelled_flag
-            return cancelled_flag
-
-        with signal_handler(conda_signal_handler), time_recorder(
-            "fetch_extract_execute"
-        ), ThreadPoolExecutor(
-            context.fetch_threads
-        ) as fetch_executor, ThreadPoolExecutor(EXTRACT_THREADS) as extract_executor:
-            for prec_or_spec, (
-                cache_action,
-                extract_action,
-            ) in self.paired_actions.items():
-                if cache_action is None and extract_action is None:
-                    # Not sure when this is reached.
-                    continue
+            if not context.verbose and not context.quiet and not context.json:
+                print(
+                    "\nDownloading and Extracting Packages:",
+                    end="\n" if IS_INTERACTIVE else " ...working...",
+                )
+            else:
+                log.debug(
+                    "prepared package cache actions:\n"
+                    "  cache_actions:\n"
+                    "    %s\n"
+                    "  extract_actions:\n"
+                    "    %s\n",
+                    "\n    ".join(str(ca) for ca in self.cache_actions),
+                    "\n    ".join(str(ea) for ea in self.extract_actions),
+                )
 
-                progress_bar = self._progress_bar(prec_or_spec, leave=False)
+            exceptions = []
+            progress_bars = {}
+            futures: list[Future] = []
 
-                progress_bars[prec_or_spec] = progress_bar
+            cancelled_flag = False
 
-                future = fetch_executor.submit(
-                    do_cache_action,
-                    prec_or_spec,
+            def cancelled():
+                """
+                Used to cancel download threads.
+                """
+                nonlocal cancelled_flag
+                return cancelled_flag
+
+            with signal_handler(conda_signal_handler), time_recorder(
```

Comment:
Some of the ugliness of this code is related to the non-threadsafety of the progress bars. For example if we try to remove one they all break. We also try to interact with the bars through the main thread. Is the new system able to address concurrency in a better way?

Reply:
My primary goal in refactoring this code was being able to accommodate defining new progress bar implementations, and I was able to accomplish the behavior you describe with my rich implementation:

- https://github.com/travishathaway/conda-rich/blob/main/conda_rich/hooks.py

I agree with wanting with wanting to improve the tqdm experience, but I also fear that this may hold this work back even further.

I propose that we add the following issue:

- https://github.com/conda/conda/issues/12210

To the follow epic:

- https://github.com/conda/conda/issues/13707

So that we don't loose track of it and ensure that the user experience is just as good with tqdm as it will be with rich.